crates/llms/src/chat/nsql/default.rs

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+     https://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use async_openai::{
+    error::OpenAIError,
+    types::{
+        ChatCompletionRequestSystemMessageArgs, CreateChatCompletionRequest,
+        CreateChatCompletionRequestArgs, CreateChatCompletionResponse, ResponseFormat,
+    },
+};
+
+use super::SqlGeneration;
+
+pub struct DefaultSqlGeneration;
+
+impl SqlGeneration for DefaultSqlGeneration {
+    fn create_request_for_query(
+        &self,
+        model_id: &str,
+        query: &str,
+        create_table_statements: &[String],
+    ) -> Result<CreateChatCompletionRequest, OpenAIError> {
+        let prompt = format!(
+            "```SQL\n{table_create_schemas}```\nTask: Write a SQL query to answer this question: _\"{query}\"_. Instruction: Return only valid SQL code, nothing additional.",
+            table_create_schemas=create_table_statements.join("\n")
+        );
+
+        CreateChatCompletionRequestArgs::default()
+            .model(model_id)
+            .response_format(ResponseFormat::Text)
+            .messages(vec![ChatCompletionRequestSystemMessageArgs::default()
+                .content(prompt)
+                .build()?
+                .into()])
+            .build()
+    }
+
+    fn parse_response(
+        &self,
+        resp: CreateChatCompletionResponse,
+    ) -> Result<Option<String>, OpenAIError> {
+        // Often models prefix with '```sql', (since markdown syntax in request). If present, remove.
+        let value = resp
+            .choices
+            .iter()
+            .find_map(|c| c.message.content.clone())
+            .map(|c| c.strip_prefix("```SQL").unwrap_or(c.as_str()).to_string());
```

Comment:
Do we have to strip the ``` suffix, too?

Reply:
Surprisingly alot of models don't. This is going to need to be some experimentation.